File: Packages/NotebarCore/Sources/NotebarCore/Panel/PanelTiming.swift
import Foundation

/// Timing defaults from spec section 4.3. Every value here becomes a user
/// setting in M4; nothing may inline these numbers at a call site.
public enum PanelTiming {
    /// Cursor must rest in the trigger zone this long before expanding.
    /// Prevents accidental opens when reaching for a scrollbar.
    public static let edgeDwell: TimeInterval = 0.120

    /// Cursor must stay outside the panel this long before collapsing.
    public static let exitDwell: TimeInterval = 0.350

    /// The collapsed handle's size. It lives here rather than in the app
    /// target's Tokens because `triggerWidth` must equal it — the handle *is*
    /// the target, and two modules disagreeing about its width is exactly how
    /// the affordance stopped working.
    public static let handleWidth: CGFloat = 30
    public static let handleHeight: CGFloat = 56

    /// The expanded panel's fixed width (spec §4.1). Flush to the right
    /// screen edge.
    public static let panelWidth: CGFloat = 340

    /// Fraction of the screen's visible height the expanded panel occupies,
    /// vertically centred (spec §4.1) — a card, not a full-height column.
    public static let panelHeightFraction: CGFloat = 0.70

    /// Fraction of `visibleFrame.width` the panel occupies when maximized
    /// (spec §6.1). At that width it fills the full screen height instead of
    /// `panelHeightFraction` — maximized reads as a docked half-screen
    /// column, not a bigger card.
    public static let maximizedWidthFraction: CGFloat = 0.5

    /// Width of the activation strip at the screen edge. Equal to the handle
    /// width so that hovering the handle — the only thing the user can see —
    /// arms the panel. The 120 ms `edgeDwell` remains the guard against
    /// accidental opens, not a narrow target.
    public static let triggerWidth: CGFloat = handleWidth

    /// Distance from the edge at which polling speeds up.
    public static let proximityWidth: CGFloat = 80

    /// Cursor must clear the panel bounds by this margin before the exit timer starts.
    public static let exitSlop: CGFloat = 24

    public static let expandDuration: TimeInterval = 0.180

    /// Deliberately faster than expanding — reads as responsive, not sluggish.
    public static let collapseDuration: TimeInterval = 0.140

    /// How long a *focused* editor keeps the panel open after the last
    /// keystroke. Longer than `typingGrace` because losing your place
    /// mid-thought is worse than a few extra seconds of panel — but finite,
    /// unlike the original design, which treated focus as absolute. That was
    /// wrong twice over: `firstResponder` is per-window and does not clear
    /// when the user looks away or the app deactivates, so a note clicked once
    /// held the panel open until the app quit. The cursor having already left
    /// the panel is itself decent evidence the user is done; to keep it open
    /// indefinitely, pin it.
    public static let focusedIdleGrace: TimeInterval = 10.0

    /// How long after the last keystroke the panel still counts as "in use".
    /// Referenced by `PanelMachine.shouldCollapse`.
    public static let typingGrace: TimeInterval = 2.0

    // MARK: - Activation settings (spec §6.5)

    /// Sane bounds for `edgeDwell` when it becomes a user setting. Used both
    /// as the slider's range in Settings and as the clamp `AppStateRepository`
    /// applies to a stored value on read, so a hand-edited database can't
    /// push it further than the UI ever could. Zero is left in: an open
    /// delay of zero is merely eager, not hostile.
    public static let edgeDwellRange: ClosedRange<TimeInterval> = 0...0.5

    /// Sane bounds for `exitDwell`. The floor is deliberately above zero —
    /// an `exitDwell` of 0 makes the panel collapse the instant the cursor
    /// leaves, which is the hostile behaviour spec §4.4 exists to prevent —
    /// and the ceiling keeps a stored value from making the panel look stuck
    /// open.
    public static let exitDwellRange: ClosedRange<TimeInterval> = 0.05...2.0

    /// Sane bounds for `exitSlop`.
    public static let exitSlopRange: ClosedRange<CGFloat> = 0...100
}

File: windows/Notebar.Core/Panel/PanelTiming.cs
namespace Notebar.Core.Panel;

/// <summary>
/// Timing and geometry defaults. Every value here is either a user setting or
/// a measurement the layout depends on; nothing may inline these numbers at a
/// call site.
/// </summary>
/// <remarks>
/// Seconds for durations, device-independent pixels for lengths. The panel
/// window works in physical pixels, so <see cref="Notebar.App"/>'s
/// PanelGeometry scales these by the target monitor's DPI — these constants
/// are never used as physical pixels directly.
/// </remarks>
public static class PanelTiming
{
    /// <summary>Cursor must rest in the trigger zone this long before expanding.
    /// Prevents accidental opens when reaching for a scrollbar.</summary>
    public const double EdgeDwell = 0.120;

    /// <summary>Cursor must stay outside the panel this long before collapsing.</summary>
    public const double ExitDwell = 0.350;

    /// <summary>The collapsed handle's size. It lives here rather than with the
    /// design tokens because <see cref="TriggerWidth"/> must equal it — the
    /// handle *is* the target, and two modules disagreeing about its width is
    /// exactly how the affordance stopped working on macOS.</summary>
    public const double HandleWidth = 30;
    public const double HandleHeight = 56;

    /// <summary>The expanded panel's fixed width, flush to the right edge.</summary>
    public const double PanelWidth = 340;

    /// <summary>Fraction of the work area's height the expanded panel occupies,
    /// vertically centred — a card, not a full-height column.</summary>
    public const double PanelHeightFraction = 0.70;

    /// <summary>Fraction of the work area's width the panel occupies when
    /// maximized. At that width it fills the full work-area height instead of
    /// <see cref="PanelHeightFraction"/> — maximized reads as a docked
    /// half-screen column, not a bigger card.</summary>
    public const double MaximizedWidthFraction = 0.5;

    /// <summary>Width of the activation strip at the screen edge. Equal to the
    /// handle width so that hovering the handle — the only thing the user can
    /// see — arms the panel. <see cref="EdgeDwell"/> remains the guard against
    /// accidental opens, not a narrow target.</summary>
    public const double TriggerWidth = HandleWidth;

    /// <summary>Distance from the edge at which polling speeds up.</summary>
    public const double ProximityWidth = 80;

    /// <summary>Cursor must clear the panel bounds by this margin before the
    /// exit timer starts.</summary>
    public const double ExitSlop = 24;

    public const double ExpandDuration = 0.180;

    /// <summary>Deliberately faster than expanding — reads as responsive,
    /// not sluggish.</summary>
    public const double CollapseDuration = 0.140;

    /// <summary>How long after the last keystroke the panel still counts as
    /// "in use". Referenced by PanelMachine.ShouldCollapse.</summary>
    public const double TypingGrace = 2.0;

    // Activation settings ranges. Used both as the slider bounds in Settings
    // and as the clamp the app-state repository applies on read, so a
    // hand-edited database cannot push the panel further than the UI ever
    // could.

    /// <summary>Zero is left in: an open delay of zero is merely eager,
    /// not hostile.</summary>
    public const double EdgeDwellMin = 0.0;
    public const double EdgeDwellMax = 0.5;

    /// <summary>The floor is deliberately above zero — an exit dwell of 0
    /// makes the panel collapse the instant the cursor leaves, which is the
    /// hostile behaviour the suppression rules exist to prevent.</summary>
    public const double ExitDwellMin = 0.05;
    public const double ExitDwellMax = 2.0;

    public const double ExitSlopMin = 0.0;
    public const double ExitSlopMax = 100.0;

    /// <summary>Clamps <paramref name="value"/> into [min, max]. Used by the
    /// app-state repository on every read of a stored timing.</summary>
    public static double Clamp(double value, double min, double max) =>
        value < min ? min : value > max ? max : value;
}
